src/js_native_api_v8.cc

```diff
@@ -270,17 +270,21 @@ class RefBase : protected Finalizer, RefTracker {
 
  protected:
   inline void Finalize(bool is_env_teardown = false) override {
+    // Force deferring behavior if the finalizer happens to delete this
+    // reference.
+    if (is_env_teardown && RefCount() > 0) _refcount = 0;
+
     if (_finalize_callback != nullptr) {
       _env->CallFinalizer(_finalize_callback, _finalize_data, _finalize_hint);
     }
 
     // this is safe because if a request to delete the reference
     // is made in the finalize_callback it will defer deletion
     // to this block and set _delete_self to true
+    _finalize_ran = true;
```

Comment:
@mhdawson removing the conditions around this assignment addresses the case where the finalizer does not delete the reference during teardown. If we didn't make this change, such a finalizer would get called twice. I added a test that aborts if the finalizer gets called twice.

In principle, I think it's good to unconditionally set `_finalize_ran` to `true`, because, well, the finalizer did run. One can argue whether, in the absence of a finalizer, we should set it to `true`, but I think it still makes sense conceptually, in that "whatever finalizer there was has now fully and completely executed".

Let's see how the tests (and especially ASAN, which I couldn't manage to run locally) run against this change. FWIW I ran the tests with gcc and clang, with and without configuring with `--debug`.